src/python/torchdistx/gossip_grad.py

```diff
@@ -0,0 +1,338 @@
+                and num_nodes is not None
+            ):
+                raise ValueError(
+                    "`local_process_group` and `num_nodes` should be provided together."
+                )
+            self.local_process_group = local_process_group
+            if num_nodes < 1:
+                raise ValueError("`num_nodes` should be equal to 1 or more.")
+            self.num_nodes = num_nodes
+
+        if self.num_nodes % 2 != 0 and self.topology == Topology.CUBE:
+            raise ValueError(
+                "Current implementation doesn't support uneven number"
+                " of nodes for CUBE topology."
+            )
+
+        super().__init__(self.local_process_group)
+        self.proc_per_node = (
+            proc_per_node
+            if proc_per_node is not None
+            else self.local_process_group.size()
+        )
+        if self.proc_per_node < 1:
+            raise ValueError("`proc_per_node` should be equal to 1 or more.")
+
+        self.master_process_group = (
+            master_process_group
+            if master_process_group is not None
+            else self._create_master_group()
+        )
+
+        self.topologies = self._generate_topologies()
+        self.cur_topology = next(self.topologies)
+
+        # For `num_nodes` != power of 2 `gossip_period` should still be an int.
+        # If we only have 1 node, `gossip_period` should be equal to 1.
+        self.gossip_period = max(1, math.ceil(math.log(self.num_nodes, 2)))
+        self.iter = 0
+
+        # Set device for `dist.batch_isend_irecv`
+        self.rank = torch.cuda.current_device()
+        torch.cuda.set_device(self.rank)
+
+        # Master worker for a current local `process_group`
+        self.master_worker = c10d._get_global_rank(self.local_process_group, 0)
+
+    def _create_master_group(self):
+        r"""
+        Creates master process group, i.e. a group of workers,
+        which communicate gradients between different nodes.
+        """
+        # Every 0th worker on every node will be assigned to a master group,
+        # i.e. if number of rocesses per node is 8, master group contains
+        # 0th, 8th, 16th, 24th, 32nd, ... ranks
+        ranks = [i * self.proc_per_node for i in range(self.num_nodes)]
+        return dist.new_group(ranks)
+
+    def _generate_topologies(self):
+        r"""
+        Creates `num_nodes` random topology shuffles and returns an infinite iterator.
+        Original topology is of the form:
+            [0*K, 1*K, ... , N*K],
+        where N is the number of nodes and K - the number of workers on each node.
+        For example, with N=4 and K=8, original topology is
+            [0, 8, 16, 24]
+
+        Workers' rank values are used instead of node values for easier peer assignment
+        in a collective communication stage.
+
+        Returns:
+            An infinite iterator over created topologies
+        """
+        random.seed(self.num_nodes * 100)
```

Comment:
users might want to control random see for debugging purposes?

Reply:
@mrshenli good point, I will make it user controllable